src/Jellyfin.Networking/Udp/UdpServer.cs

```diff
@@ -0,0 +1,136 @@
+
+namespace Jellyfin.Networking.Udp;
+
+/// <summary>
+/// Provides a Udp Server.
+/// </summary>
+public sealed class UdpServer : IDisposable
+{
+    /// <summary>
+    /// The _logger.
+    /// </summary>
+    private readonly ILogger _logger;
+    private readonly IServerApplicationHost _appHost;
+    private readonly IConfiguration _config;
+
+    private readonly byte[] _receiveBuffer = new byte[8192];
+
+    private readonly Socket _udpSocket;
+    private readonly IPEndPoint _endpoint;
+    private bool _disposed;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="UdpServer" /> class.
+    /// </summary>
+    /// <param name="logger">The logger.</param>
+    /// <param name="appHost">The application host.</param>
+    /// <param name="configuration">The configuration manager.</param>
+    /// <param name="bindAddress"> The bind address.</param>
+    /// <param name="port">The port.</param>
+    public UdpServer(
+        ILogger logger,
+        IServerApplicationHost appHost,
+        IConfiguration configuration,
+        IPAddress bindAddress,
+        int port)
+    {
+        _logger = logger;
+        _appHost = appHost;
+        _config = configuration;
+
+        _endpoint = new IPEndPoint(bindAddress, port);
+
+        _udpSocket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp)
+        {
+            MulticastLoopback = false,
+        };
+        _udpSocket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+    }
+
+    private async Task RespondToV2Message(EndPoint endpoint, CancellationToken cancellationToken)
+    {
+        string? localUrl = _config[AddressOverrideKey];
+        if (string.IsNullOrEmpty(localUrl))
+        {
+            localUrl = _appHost.GetSmartApiUrl(((IPEndPoint)endpoint).Address);
+        }
+
+        if (string.IsNullOrEmpty(localUrl))
+        {
+            _logger.LogWarning("Unable to respond to server discovery request because the local ip address could not be determined.");
+            return;
+        }
+
+        var response = new ServerDiscoveryInfo(localUrl, _appHost.SystemId, _appHost.FriendlyName);
+
+        try
+        {
+            _logger.LogDebug("Sending AutoDiscovery response");
+            await _udpSocket.SendToAsync(JsonSerializer.SerializeToUtf8Bytes(response), SocketFlags.None, endpoint, cancellationToken).ConfigureAwait(false);
+        }
+        catch (SocketException ex)
+        {
+            _logger.LogError(ex, "Error sending response message");
+        }
+    }
+
+    /// <summary>
+    /// Starts the specified port.
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+    public void Start(CancellationToken cancellationToken)
+    {
+        _udpSocket.Bind(_endpoint);
+
+        _ = Task.Run(async () => await BeginReceiveAsync(cancellationToken).ConfigureAwait(false), cancellationToken).ConfigureAwait(false);
```

Comment:
I would prefer to use `Task.Factory.StartNew(..., TaskCreationOptions.LongRunning)` to not clog up the scheduler

Reply:
That is what `TaskCreationOptions.LongRunning` does. It does not schedule the Task on the local Thread pool thread.

Reply:
Only the first part of the code will be on that longrunning thread. Thankfully, it's very easy to verify :)

```
            var task =  Task.Factory.StartNew(async () =>
            {
                while (true)
                {
                    Console.WriteLine($"Before await. Is thread pool thread? {Thread.CurrentThread.IsThreadPoolThread}");
                    await Task.Delay(100);
                    Console.WriteLine($"After await. Is thread pool thread? {Thread.CurrentThread.IsThreadPoolThread}");
                }
            }, TaskCreationOptions.LongRunning);
            Console.ReadKey();
```

Output:

```
Before await. Is thread pool thread? False
After await. Is thread pool thread? True
Before await. Is thread pool thread? True
After await. Is thread pool thread? True
Before await. Is thread pool thread? True
After await. Is thread pool thread? True
```

Reply:
That is correct, but when you use async-await it's not long running, and as you can see by the sample code I provided, the continuation is now bound to the Thread pool.